File: deprecated/mod_inp.py
MODULE MOD_INP

  CONTAINS

    SUBROUTINE GET_VAL(LNUM, NUMCHAR, TEXT_LINE, VARNAME, VARTYPE, LOGVAL, STRINGVAL, REALVAL, INTVAL, NVAL)
    ! Decompose input line into variable name and value
      USE MOD_PREC
      IMPLICIT NONE
      
      INTRINSIC ADJUSTL
      
      INTEGER, INTENT(IN) :: LNUM,NUMCHAR
      CHARACTER(LEN = NUMCHAR) :: TEXT_LINE
      CHARACTER(LEN = 20), INTENT(OUT) :: VARNAME
      CHARACTER(LEN = 7), INTENT(OUT) :: VARTYPE
      LOGICAL, INTENT(OUT) :: LOGVAL
      CHARACTER(LEN = 80), INTENT(OUT) :: STRINGVAL(150)
      REAL(SP), INTENT(INOUT) :: REALVAL(150)
      INTEGER, INTENT(INOUT) :: INTVAL(150)
      INTEGER, INTENT(OUT) :: NVAL

      CHARACTER(LEN = NUMCHAR) :: VARVAL,TEMP,FRAG(200)
      CHARACTER(LEN = 80) :: TSTRING
      CHARACTER(LEN = 6) :: ERRSTRING
      CHARACTER(LEN = 16) :: NUMCHARS 
      INTEGER :: LENGTH, EQLOC, LVARVAL, DOTLOC, ii, J, LOCEX, NP
      LOGICAL :: ONFRAG

      FRAG = " "
      NUMCHARS = "0123456789+-Ee. " 
      VARTYPE = "error"
      LOGVAL = .FALSE.
      LENGTH = LEN_TRIM(TEXT_LINE) 
      WRITE(ERRSTRING,"(I6)") LNUM
      LOCEX = INDEX(TEXT_LINE,"!")

      ! CHECK FOR BLANK LINE OR COMMENT
      IF((LENGTH .eq. 0) .OR. (LOCEX .eq. 1))THEN
        VARTYPE = "no data"
        VARNAME = "no data"
        RETURN
      END IF

      ! CHANGE COMMAS TO BLANKS
      do ii = 1, LENGTH
        if (TEXT_LINE(ii:ii) == ",") TEXT_LINE(ii:ii) = " "
      end do

    ! REMOVING TRAILING COMMENTS
      IF (LOCEX .ne. 0) THEN
        TEMP = TEXT_LINE(1:LOCEX-1)
        TEXT_LINE = TEMP
      END IF

      ! ENSURE "=" EXISTS AND DETERMINE LOCATION
      EQLOC = INDEX(TEXT_LINE,"=")
      IF(EQLOC .eq. 0) CALL PERROR(6,'DATA LINE '//ERRSTRING//' MUST CONTAIN "=" ')

      ! SPLIT OFF VARNAME AND VARVAL STRINGS
      VARNAME = TEXT_LINE(1:EQLOC-1)
      VARVAL  = ADJUSTL(TEXT_LINE(EQLOC+1:LENGTH))
      LVARVAL = LEN_TRIM(VARVAL)
      IF(LVARVAL .eq. 0) CALL PERROR(6,'IN DATA PARAMETER FILE', 'VARIABLE LINE'//ERRSTRING//' HAS NO ASSOCIATED VALUE')

      ! DETERMINE TYPE OF VARVAL
      ! CHECK FOR LOGICAL
      IF(((VARVAL(1:1) .eq. "T") .OR. (VARVAL(1:1) .eq. "F")) .and. (LVARVAL .eq. 1))THEN 
        VARTYPE = "logical"
        IF(VARVAL(1:1) .eq. "T") LOGVAL = .TRUE.
        RETURN
      END IF

      ! CHECK IF IT IS A STRING  (CONTAINS CHARACTERS OTHER THAN 0-9,+,-,e,E,.)
      DO ii = 1, LVARVAL
        IF(INDEX(NUMCHARS,VARVAL(ii:ii)) == 0) VARTYPE = "string" 
      END DO


      ! PROCESS STRING (MAY BE MULTIPLE)
      IF(VARTYPE .eq. "string") THEN
        TSTRING = VARVAL
        STRINGVAL(1) = TSTRING 
        NVAL = 1
        ONFRAG = .TRUE.
        DO ii = 1, LVARVAL
          IF(VARVAL(ii:ii) .ne. " ")THEN
            FRAG(NVAL) = TRIM(FRAG(NVAL))//VARVAL(ii:ii)
            ONFRAG = .TRUE.
          ELSE
            IF(ONFRAG) NVAL = NVAL + 1
            ONFRAG = .FALSE.
          END IF
        END DO
        DO ii = 1, NVAL
          STRINGVAL(ii + 1) = TRIM(FRAG(ii))
        END DO
        RETURN
      END IF

      ! CHECK IF IT IS A FLOAT
      DOTLOC = INDEX(VARVAL,".")
      IF(DOTLOC .ne. 0) THEN
        VARTYPE = "float"
      ELSE
        VARTYPE = "integer"
      END IF
   
    ! FRAGMENT INTO STRINGS FOR MULTIPLE VALUES
      NP = 1
      ONFRAG = .TRUE.
      DO ii = 1, LVARVAL
        IF (VARVAL(ii:ii) .ne. " ")THEN 
          FRAG(NP) = TRIM(FRAG(NP))//VARVAL(ii:ii)
          ONFRAG = .TRUE.
        ELSE
          IF(ONFRAG) NP = NP + 1
          ONFRAG = .FALSE.
        END IF
      END DO

      ! EXTRACT NUMBER(S) FROM CHARACTER STRINGS
      NVAL = NP
      DO ii = 1, NP
        TEMP = TRIM(FRAG(ii))
        IF (VARTYPE .eq. "float") THEN 
          READ(TEMP, *) REALVAL(ii)
        ELSE
          READ(TEMP, *) INTVAL(ii)
        END IF
      END DO
  
    END SUBROUTINE GET_VAL 


  !==============================================================================|
  integer function SCAN_FILE(FNAME,VNAME,ISCAL,FSCAL,IVEC,FVEC,CVEC,NSZE,CVAL,LVAL)
  !   Scan an Input File for a Variable
  !   RETURN VALUE:
  !        0 = FILE FOUND, VARIABLE VALUE FOUND
  !       -1 = FILE DOES NOT EXIST OR PERMISSIONS ARE INCORRECT
  !       -2 = VARIABLE NOT FOUND OR IMPROPERLY SET
  !       -3 = VARIABLE IS OF DIFFERENT TYPE, CHECK INPUT FILE
  !       -4 = VECTOR PROVIDED BUT DATA IS SCALAR TYPE
  !       -5 = NO DATATYPE DESIRED, EXITING

  !   REQUIRED INPUT:
  !        FNAME = File Name
  !        FSIZE = Length of Filename

  !   OPTIONAL (MUST PROVIDE ONE)
  !        ISCAL = INTEGER SCALAR
  !        FSCAL = FLOAT SCALAR
  !        CVAL = CHARACTER VARIABLE
  !        LVAL = LOGICAL VARIABLE
  !        IVEC = INTEGER VECTOR **
  !        FVEC = FLOAT VECTOR **
  !        CVEC = STRING VECTOR **
  !      **NSZE = ARRAY SIZE (MUST BE PROVIDED WITH IVEC/FVEC)

     USE MOD_PREC
     IMPLICIT NONE
     CHARACTER(LEN = *) :: FNAME, VNAME
     INTEGER, INTENT(INOUT), OPTIONAL :: ISCAL, IVEC(*)
     REAL(SP), INTENT(INOUT), OPTIONAL :: FSCAL, FVEC(*)
     CHARACTER(LEN=80), OPTIONAL :: CVAL, CVEC(*)
     LOGICAL, INTENT(INOUT), OPTIONAL :: LVAL
     INTEGER, INTENT(INOUT), OPTIONAL :: NSZE
   
     REAL(SP) REALVAL(150)
     INTEGER  INTVAL(150)
     CHARACTER(LEN=20 ) :: VARNAME
     CHARACTER(LEN=80 ) :: STRINGVAL(150),TITLE
     CHARACTER(LEN=80 ) :: INPLINE
     CHARACTER(LEN=400) :: TLINE
     CHARACTER(LEN=7  ) :: VARTYPE
     CHARACTER(LEN=20 ), DIMENSION(200) :: SET
     INTEGER :: I, NVAL, J, NSET, NLINE, NREP
     LOGICAL :: SETYES, ALLSET, CHECK, LOGVAL


     SCAN_FILE = 0

    ! OPEN THE INPUT FILE
    INQUIRE(FILE = TRIM(FNAME), EXIST = CHECK)
    IF (.NOT. CHECK) THEN
      SCAN_FILE = -1
      RETURN
    END IF

    OPEN(10, FILE=TRIM(FNAME)); REWIND(10) 


  ! SCAN THE FILE FOR THE VARIABLE NAME
     NSET = 0
     NLINE = 0
     DO WHILE (.TRUE.)
       TLINE(1:LEN(TLINE)) = ' ' 
       NREP  = 0
       NLINE = NLINE + 1
       READ(10,'(a)',END=20) INPLINE
       TLINE(1:80) = INPLINE(1:80)

  ! PROCESS LINE CONTINUATIONS
   110 CONTINUE
       I = LEN_TRIM(INPLINE)
       IF (I .ne. 0) THEN
  !#    if defined (COMPAQ) || defined (INTEL) || defined (IRIX)
       !IF( INPLINE(I-1:I) == '\\')THEN
  !#    else
       IF ( INPLINE(I-1:I) == '\\\\') THEN
  !#    endif
         NREP = NREP + 1
         READ(10,'(a)',END=20) INPLINE
         NLINE = NLINE + 1
         TLINE( NREP*80 + 1 : NREP*80 +80) = INPLINE(1:80)
         GOTO 110
       END IF
       END IF
       IF(NREP > 4)CALL PERROR(6,"CANNOT HAVE > 4 LINE CONTINUATIONS")

    ! REMOVE LINE CONTINUATION CHARACTER \\
       IF (NREP .gt. 0)THEN
         DO I = 2, LEN_TRIM(TLINE)
  !#    if defined (COMPAQ) || defined (INTEL) || defined (IRIX)
           !IF( TLINE(I-1:I) == '\\') TLINE(I-1:I) = '  '
  !#        else
           IF( TLINE(I-1:I) == '\\\\') TLINE(I-1:I) = '  '
  !#        endif
         END DO
       END IF
       
    ! PROCESS THE LINE
    CALL GET_VAL(NLINE, LEN_TRIM(TLINE), ADJUSTL(TLINE), VARNAME, VARTYPE, LOGVAL, STRINGVAL, REALVAL, INTVAL, NVAL)

  ! IF VARNAME MATCHES, PROCESS VARIABLE AND ERROR-CHECK

       IF(TRIM(VARNAME) .eq. TRIM(VNAME))THEN

         IF(PRESENT(ISCAL))THEN
           IF(VARTYPE .eq. 'integer')THEN
             ISCAL = INTVAL(1)
             RETURN
           ELSE
             SCAN_FILE = -3
           END IF
         ELSE IF(PRESENT(FSCAL))THEN
           IF(VARTYPE .eq. 'float')THEN
             FSCAL = REALVAL(1)
             RETURN
           ELSE
             SCAN_FILE = -3
           END IF
         ELSE IF(PRESENT(CVAL))THEN
           IF(VARTYPE .eq. 'string')THEN
             CVAL = STRINGVAL(1) 
             RETURN
           ELSE
             SCAN_FILE = -3
           END IF
         ELSE IF(PRESENT(LVAL))THEN
           IF(VARTYPE .eq. 'logical')THEN
             LVAL = LOGVAL 
             RETURN
           ELSE
             SCAN_FILE = -3
           END IF
         ELSE IF (PRESENT(IVEC)) THEN
           IF (NVAL .gt. 1) THEN
             IF (VARTYPE .eq. 'integer') THEN
               IVEC(1:NVAL) = INTVAL(1:NVAL) 
               NSZE = NVAL
               RETURN
             ELSE
               SCAN_FILE = -3
             END IF
             ELSE
             SCAN_FILE = -4 
           END IF
         ELSE IF (PRESENT(FVEC)) THEN
           IF (NVAL .gt. 1) THEN
             IF (VARTYPE .eq. 'float') THEN
               FVEC(1:NVAL) = REALVAL(1:NVAL) 
               NSZE = NVAL           
               RETURN
             ELSE
               SCAN_FILE = -3
             END IF
           ELSE
             SCAN_FILE = -4 
           END IF
         ELSE IF (PRESENT(CVEC)) THEN
           IF(NVAL .gt. 0)THEN
             IF (VARTYPE .eq. 'string') THEN
               CVEC(1:NVAL) = STRINGVAL(2:NVAL+1)
               NSZE = NVAL 
               RETURN
             ELSE
               SCAN_FILE = -3
             END IF
           ELSE
             SCAN_FILE = -4
           END IF
         ELSE
           SCAN_FILE = -5
         END IF
       END IF  !!VARIABLE IS CORRECT
            
     END DO !!LOOP OVER INPUT FILE
   20 CLOSE(10) 
     SCAN_FILE = -2

     RETURN 
     END FUNCTION SCAN_FILE


    SUBROUTINE PERROR(IOUT, ER1, ER2, ER3, ER4) !--Huang change -- add from FVCOM "mod_utils.F"

       IMPLICIT NONE
       INTEGER :: IOUT
       CHARACTER(LEN = *)             :: ER1
       CHARACTER(LEN = *), OPTIONAL   :: ER2
       CHARACTER(LEN = *), OPTIONAL   :: ER3
       CHARACTER(LEN = *), OPTIONAL   :: ER4

       WRITE(IOUT, *) '==================ERROR=================================='
       WRITE(IOUT, *) ER1
       IF (PRESENT(ER2)) WRITE(IOUT, *) ER2
       IF (PRESENT(ER3)) WRITE(IOUT, *) ER3
       IF (PRESENT(ER4)) WRITE(IOUT, *) ER4
       WRITE(IOUT, *) '========================================================='
       STOP
       RETURN

    END SUBROUTINE PERROR

END MODULE MOD_INP

File: mod_inp.f90
module MOD_INP

contains

    subroutine DATA_RUN
        ! Input Parameters Which Control the Model Run
        use ALL_VARS

        implicit none

        integer :: ISCAN
        character(len = 120) :: filename
        integer :: ii


        ! Read in variables and set values
        filename = "./"//trim(folderprefix)//"/"//trim(CASENAME)//"_run.dat"

        ! Info file
        ISCAN = scan(filename, "INFOFILE", CVAL = INFOFILE)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING INFOFILE: ', ISCAN
            stop
        end if

        ! Open runtime info file
        IPT = 71
        if (trim(INFOFILE) /= "screen") then
            open(IPT, FILE = trim(INFOFILE))
        else
            IPT = 6
        end if

        ! External time step (DTI)
        ISCAN = scan(trim(filename), "DTI", FSCAL = DTI)
        if (ISCAN /= 0) then
            write (IPT, *) 'ERROR READING DTI: ', ISCAN
            stop
        end if

        ! Input time step of flow fields (instp)
        ISCAN = scan(trim(filename),"INSTP", FSCAL = INSTP)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING INSTP: ', ISCAN
            stop
        end if

        ! External time step (DTOUT)
        ISCAN = scan(trim(filename), "DTOUT", FSCAL = DTOUT)
        if (ISCAN /= 0) then
            write (IPT, *) 'ERROR READING DTOUT: ', ISCAN
            stop
        end if

        ! Total time to move drifters (TDRIFT)
        ISCAN = scan(trim(filename), "TDRIFT", ISCAL = TDRIFT)
        if (ISCAN /= 0)then
            write(IPT, *) 'ERROR READING TDRIFT: ', ISCAN
            stop
        end if


        ! Input year of run (YEARLAG)
        ISCAN = scan(trim( filename),"YEARLAG",ISCAL = YEARLAG)
        if (ISCAN /= 0) then
            write(IPT, *)'ERROR READING YEARLAG: ', ISCAN
            stop
        end if

        ! Input month of run (MONTHLAG)
        ISCAN = scan(trim(filename), "MONTHLAG", ISCAL = MONTHLAG)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING YEARLAG: ', ISCAN
            stop
        end if

        ! Input day of run (DAYLAG)
        ISCAN = scan(trim(filename), "DAYLAG", ISCAL = DAYLAG)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING DAYLAG: ', ISCAN
            stop
        end if

        ! Input hour of run (HOURLAG)
        ISCAN = scan(trim(filename), "HOURLAG", ISCAL = HOURLAG)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING HOURLAG: ', ISCAN
            stop
        end if

        ! "P_SIGMA" turns on vertical location of particles in sigma
        ISCAN = scan(trim(filename), "P_SIGMA", LVAL = P_SIGMA)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING P_SIGMA: ', ISCAN
            if (ISCAN == -2) then
                write(IPT,*)'VARIABLE NOT FOUND IN INPUT FILE: ',trim(filename)
            end if
            stop
        end if

        ! "OUT_SIGMA" TURNS ON VERTICAL LOCATION OF PARTICLES IN SIGMA
        ISCAN = scan(trim(filename), "OUT_SIGMA", LVAL = OUT_SIGMA)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING OUT_SIGMA: ', ISCAN
            if (ISCAN == -2) then
                write(IPT, *) 'VARIABLE NOT FOUND IN INPUT FILE: ', trim(filename)
            end if
            stop
        end if


        ! "F_DEPTH" KEEP SAME Z DEPTH ALONG THE TRACKING
        ISCAN = scan(trim(filename), "F_DEPTH", LVAL = F_DEPTH)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING F_DEPTH: ', ISCAN
            if (ISCAN == -2) then
                write(IPT, *) 'VARIABLE NOT FOUND IN INPUT FILE: ', trim(filename)
            end if
            stop
        end if

        ! RANDOM WALK CHOICE
        ISCAN = scan(trim(filename), "IRW", ISCAL = IRW)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING IRW: ', ISCAN
            stop
        end if

        ! Horizontal diffusion coefficient (DHOR)
        ISCAN = scan(trim(filename), "DHOR", FSCAL = DHOR)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING DHOR: ', ISCAN
            stop
        end if

        !  RANDOM WALK TIME STEP (DTRW)
        ISCAN = scan(trim(filename), "DTRW", FSCAL = DTRW)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING DTRW: ', ISCAN
            stop
        end if

        ! "GEOAREA" DIRECTORY FOR INPUT FILES
        ISCAN = scan(filename,"GEOAREA",CVAL = GEOAREA)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING GEOAREA: ', ISCAN
            stop
        end if

        ii = len_trim(GEOAREA)
        if(GEOAREA(ii:ii) == "/") GEOAREA(ii:ii) = " "


        ! DIRECTORY FOR INPUT FILES
        ISCAN = scan(filename, "INPDIR", CVAL = INPDIR)
        if (ISCAN /= 0) then
            write(IPT, *)'ERROR READING INPDIR: ', ISCAN
            stop
        end if

        ii = len_trim(INPDIR)
        if(INPDIR(ii:ii) == "/") INPDIR(ii:ii) = " "


        ! INPUT FILES
        ISCAN = scan(filename, "LAGINI", CVAL = LAGINI)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING LAGINI: ', ISCAN
            stop
        end if

        ii = len_trim(LAGINI)
        if (LAGINI(ii:ii) == "/") LAGINI(ii:ii) = " "


        ! "OUTDIR"
        ISCAN = scan(filename, "OUTDIR", CVAL = OUTDIR)
        if (ISCAN /= 0) then
            write(IPT, *) 'ERROR READING OUTDIR: ', ISCAN
            stop
        end if

        ii = len_trim(OUTDIR)
        if (OUTDIR(ii:ii) == "/") OUTDIR(ii:ii) = " "

        ! Set unit values for input output files
        IOPAR=11
        INLAG=13

    end subroutine

    subroutine parse_line(LNUM, NUMCHAR, key, type_of, LOGVAL, STRINGVAL, REALVAL, INTVAL, NVAL)
        ! Decompose input line into variable name and value
        use parameters
        implicit none

        intrinsic adjustl

        integer, intent(in) :: LNUM, NUMCHAR
        character(len = NUMCHAR) :: text
        character(len = 20), intent(out) :: key
        character(len = 7), intent(out) :: type_of

        logical, intent(out) :: LOGVAL
        character(LEN = 80), intent(out), dimension(150) :: STRINGVAL
        real(SP), intent(inout), dimension(150) :: REALVAL
        integer, intent(inout), dimension(150) :: INTVAL
        integer, intent(out) :: NVAL

        character(len = NUMCHAR) :: value, TEMP, fragments(200)
        character(len = 80) :: TSTRING
        character(len = 6) :: ERRSTRING
        character(len = 16) :: NUMCHARS
        integer :: EQLOC, length, ii, LOCEX, NP
        logical :: flag

        fragments = " "
        NUMCHARS = "0123456789+-Ee. "
        type_of = "error"
        LOGVAL = .false.
        
        write(ERRSTRING, "(I6)") LNUM
        
        LOCEX = index(text, "!")
        if (LOCEX /= 0) text = text(1:LOCEX-1)
        length = len_trim(text)
        
        if (length == 0) then
            type_of = "none"
            key = "none"
            return
        end if

        ! Commas to spaces
        where (text(:) == ",")
            text(:) = " "
        end where
        
        ! Find assignment "="
        EQLOC = INDEX(text,"=")
        IF (EQLOC == 0) CALL raise(6,'DATA LINE '//ERRSTRING//' MUST CONTAIN "=" ')

        ! split name and value substrings
        key = text(1:EQLOC-1)
        value  = adjustl(text(EQLOC+1:LENGTH))
        length = len_trim(value)
        
        IF (length == 0) CALL raise(6,'IN DATA PARAMETER FILE', 'VARIABLE LINE'//ERRSTRING//' HAS NO ASSOCIATED VALUE')

        ! DETERMINE TYPE OF value
        ! CHECK FOR LOGICAL
        IF (((value(1:1) == "T") .or. (value(1:1) == "F")) .and. (length == 1)) then
            type_of = "logical"
            if (value(1:1) == "T") LOGVAL = .true.
            return
        end if

        ! CHECK IF IT IS A STRING  (CONTAINS characterS OTHER THAN 0-9,+,-,e,E,.)
        do ii = 1, length
            if (INDEX(NUMCHARS, value(ii:ii)) == 0) type_of = "string"
        end do

        ! PROCESS STRING (MAY BE MULTIPLE)
        if (type_of == "string") then
            TSTRING = value
            stringval(1) = TSTRING
            NVAL = 1
            flag = .true.
            do ii = 1, length
                if (value(ii:ii) /= " ") then 
                    fragments(NVAL) = trim(fragments(NVAL)) // value(ii:ii)
                    flag = .true.
                else
                    if (flag) NVAL = NVAL + 1
                    flag = .false.
                end if
            end do
            
            do ii = 1, NVAL
                stringval(ii + 1) = trim(fragments(ii))
            end do
            return
            
        end if

        type_of = merge("float", "integer", index(value, ".") /= 0)
        
        ! Split lines
        NP = 1
        flag = .true.
        do ii = 1, length
            if (value(ii:ii) /= " ") then
                fragments(NP) = trim(fragments(NP)) // value(ii:ii)
                flag = .true.
            else
                if (flag) NP = NP + 1
                flag = .false.
            end if
        end do

        ! numerical
        NVAL = NP
        do ii = 1, NP
            read(trim(fragments(ii)), *) merge(realval(ii), intval(ii), type_of == "float")
        end do

    end subroutine
    

    integer function scan(FNAME, VNAME, ISCAL, FSCAL, IVEC, FVEC, CVEC, NSZE, CVAL, LVAL)
        !   Scan an Input File for a Variable
        !   RETURN VALUE:
        !        0 = FILE FOUND, VARIABLE VALUE FOUND
        !       -1 = FILE DOES NOT EXIST OR PERMISSIONS ARE INCORRECT
        !       -2 = VARIABLE NOT FOUND OR IMPROPERLY SET
        !       -3 = VARIABLE IS OF DIFFERENT TYPE, CHECK INPUT FILE
        !       -4 = VECTOR PROVIDED BUT DATA IS SCALAR TYPE
        !       -5 = NO DATATYPE DESIRED, EXITING

        !   REQUIRED INPUT:
        !        FNAME = File Name
        !        FSIZE = Length of Filename

        !   optional (MUST PROVIDE ONE)
        !        ISCAL = integer SCALAR
        !        FSCAL = FLOAT SCALAR
        !        CVAL = character VARIABLE
        !        LVAL = LOGICAL VARIABLE
        !        IVEC = integer VECTOR **
        !        FVEC = FLOAT VECTOR **
        !        CVEC = STRING VECTOR **
        !      **NSZE = ARRAY SIZE (MUST BE PROVIDED WITH IVEC/FVEC)

        use parameters
        implicit none
        character(LEN = *) :: FNAME, VNAME
        integer, intent(inout), optional :: ISCAL, IVEC(*)
        REAL(SP), intent(inout), optional :: FSCAL, FVEC(*)
        character(LEN=80), optional :: CVAL, CVEC(*)
        LOGICAL, intent(inout), optional :: LVAL
        integer, intent(inout), optional :: NSZE

        REAL(SP) REALVAL(150)
        integer  INTVAL(150)
        character(LEN=20 ) :: name
        character(LEN=80 ) :: STRINGVAL(150),TITLE
        character(LEN=80 ) :: line
        character(LEN=400) :: copy
        character(LEN=7  ) :: type_of
        character(LEN=20 ), DIMENSION(200) :: SET
        integer :: last, NVAL, lines, NREP
        logical :: SETYES, ALLSET, CHECK, LOGVAL
        character(len=*), parameter :: continue_line = "////"


        scan = 0

        ! OPEN THE INPUT FILE
        inquire(file=TRIM(FNAME), exist=CHECK)
        if (.not. CHECK) then
            scan = -1
        end if

        open(10, file=trim(FNAME))
        rewind(10)
        
        lines = 0
        do while (.true.)

            copy(1:len(copy)) = ' '
            NREP  = 0
            lines = lines + 1
            read(10,'(a)', end=20) line
            copy(1:80) = line(1:80)

            ! PROCESS LINE CONTINUATIONS
            110 CONTINUE
            last = len_trim(line)
            if (last /= 0) then
                if ( line(last-1:last) == '\\\\') then
                   
                    NREP = NREP + 1
                    read(10, '(a)', end=20) line
                    lines = lines + 1
                    copy( NREP*80 + 1 : NREP*80 +80) = line(1:80)
                    GOTO 110
                end if
            end if

            ! REMOVE LINE CONTINUATION character \\
            if (NREP > 0) then
                do last = 2, LEN_TRIM(copy)
                    if ( copy(last-1:last) == '\\\\') copy(last-1:last) = '  '
                end do
            end if

            call parse_line(lines, len_trim(copy), adjustl(copy), name, type_of, LOGVAL, STRINGVAL, REALVAL, INTVAL, NVAL)

            ! IF name MATCHES, PROCESS VARIABLE AND ERROR-CHECK

            if (trim(name) == trim(VNAME)) then

                if (PRESENT(ISCAL)) then
                    if (type_of == 'integer') then
                        ISCAL = INTVAL(1)
                        return
                    else
                        scan = -3
                    end if
                elseif(present(FSCAL)) then
                    if (type_of == 'float') then
                        FSCAL = REALVAL(1)
                        return
                    else
                        scan = -3
                    end if
                elseif(present(CVAL))THEN
                    if (type_of == 'string') then
                        CVAL = STRINGVAL(1)
                        return
                    else
                        scan = -3
                    end if
                elseif (present(LVAL)) THEN
                    if (type_of == 'logical') then
                        LVAL = LOGVAL
                        return
                    else
                        scan = -3
                    end if
                else if (present(IVEC)) then
                    if (NVAL > 1) then
                        if (type_of == 'integer') then
                            IVEC(1:NVAL) = INTVAL(1:NVAL)
                            NSZE = NVAL
                            return
                        else
                            scan = -3
                        end if
                    else
                        scan = -4
                    end if
                elseif (present(FVEC)) then
                    if (NVAL > 1) then
                        IF (type_of == 'float') then
                            FVEC(1:NVAL) = REALVAL(1:NVAL)
                            NSZE = NVAL
                            return
                        else
                            scan = -3
                        end if
                    else
                        scan = -4
                    end if
                elseif (present(CVEC)) then
                    if (NVAL > 0) then
                        if (type_of == 'string') then
                            CVEC(1:NVAL) = STRINGVAL(2:NVAL+1)
                            NSZE = NVAL
                            return
                        else
                            scan = -3
                        end if
                    else
                        scan = -4
                    end if
                else
                    scan = -5
                end if
            end if
        end do
        20 close(10)
        scan = -2
    end function

end module